cpp/src/cluster/detail/single_linkage.cuh

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2021-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "agglomerative.cuh"
+#include "connectivities.cuh"
+#include "mst.cuh"
+#include <cuvs/cluster/agglomerative.hpp>
+#include <raft/core/resource/cuda_stream.hpp>
+#include <raft/util/cudart_utils.hpp>
+
+#include <rmm/device_uvector.hpp>
+
+namespace cuvs::cluster::agglomerative::detail {
+
+static const size_t EMPTY = 0;
+
+/**
+ * Single-linkage clustering, capable of constructing a KNN graph to
+ * scale the algorithm beyond the n^2 memory consumption of implementations
+ * that use the fully-connected graph of pairwise distances by connecting
+ * a knn graph when k is not large enough to connect it.
+
+ * @tparam value_idx
+ * @tparam value_t
+ * @tparam dist_type method to use for constructing connectivities graph
+ * @param[in] handle raft handle
+ * @param[in] X dense input matrix in row-major layout
+ * @param[in] m number of rows in X
+ * @param[in] n number of columns in X
+ * @param[in] metric distance metrix to use when constructing connectivities graph
+ * @param[out] out struct containing output dendrogram and cluster assignments
+ * @param[in] c a constant used when constructing connectivities from knn graph. Allows the indirect
+ control
+ *            of k. The algorithm will set `k = log(n) + c`
+ * @param[in] n_clusters number of clusters to assign data samples
+ */
+template <typename value_idx, typename value_t, Linkage dist_type>
```

Comment:
I have followed the stack up and down, and I don't think `Linkage` needs to be a template parameter, we can get away with it as a function argument and reduce the number of instantiations. While I don't think we are invoking any kernels twice so the savings might not be extraordinary, I think it's worth reducing number of templates anyway.

I'm okay with doing this as a follow-on if you consider this worth refactoring later on.

Reply:
There was a strict reason behind this, actually. If you recall, initially we had half of single linkage in cuml and half in raft. We had the pairwise distance primitive in raft but the knn primitives hadn't been moved over yet. This is essentially doing an extensible strategy pattern that allowed the raft version to plug into the cuml version. 

I agree it could go away now. I don't think this migration is the place to do that, though. I'm trying to change as little in the internals as possible so that we have a baseline point for the move. The goal is to have a heavy focus on continuing to improve these internals over the next few versions.

I created a blanket issue for removing the unneeded templates. I'll add this one to that issue (and please also add things as you see / think of them just so we don't lose sight of them.